allow users to be externally redirected after logging in

ericbuckley opened the pull request:
closes department-of-veterans-affairs/va.gov-team#10630

The standalone sign-in page is required to accept parameters to redirect to external applications once login is successful. The first pass of this added was passing along `application` and `to` query parameters to `vets-api` and letting it determine redirect logic. 

This added unnecessary complexity to a lot of the helper authentication methods. Instead, this PR isolates redirect logic to a 
single utility within the login callback, successfully establishing a va.gov session before redirecting to an external site like `myvahealth`

ericbuckley committed:
adding a function to calculate the standalone external redirect based on query params

ericbuckley committed:
no need to pass along application and to params, FE will now redirect the user to an external app rather than the BE

hinzed1127 commented on src/platform/user/authentication/utilities.js:
```diff
@@ -79,12 +72,22 @@ function redirectWithGAClientId(redirectUrl) {
 }
 
+function standaloneRedirect() {
+  const { application, to } = window.location.query;
```
I believe `location.query` is specific to the react router and not a part of the standard [browser API](https://developer.mozilla.org/en-US/docs/Web/API/Location). I'm going to update this to use `URLSearchParams` along with tests

Commit:
Use URLSearchParams for redirect construction, add tests

ericbuckley committed:
Merge branch 'master' into eb-external-redirects

ericbuckley committed:
adding one more test case

ericbuckley committed:
fixing test cases

ericbuckley committed:
Merge branch 'master' into eb-external-redirects

ncksllvn commented on src/platform/user/authentication/utilities.js:
```diff
@@ -79,12 +72,25 @@ function redirectWithGAClientId(redirectUrl) {
+export function standaloneRedirect() {
+  const searchParams = new URLSearchParams(window.location.search);
+  const application = searchParams.get('application');
+  const to = searchParams.get('to');
```
I asked this before and it was fine but just confirming that the `to` parameter is validated in some way so that we can't just put anything there - `va.gov?to=malicious-website.com`

ericbuckley replied:
if no `application` query parameter was provided, the if block on line 81 would fail and null would be returned.

ncksllvn replied:
Hm what about `a.gov?application=something&to=malicious-website.com`?

ericbuckley replied:
In that case the url would be `https://ehrm-va-test.patientportal.us.healtheintent.com/malicious-website.com`

ericbuckley committed:
Merge branch 'master' into eb-external-redirects

ericbuckley committed:
Merge branch 'master' into eb-external-redirects

patrickvinograd commented on src/platform/user/authentication/utilities.js:
```diff
@@ -75,12 +68,25 @@ function redirectWithGAClientId(redirectUrl) {
   const returnUrl =
     window.location.pathname === '/sign-in/'
-      ? window.location.origin
+      ? standaloneRedirect()
```
If `standaloneRedirect` returns null (e.g. if no application param is passed, or an un-configured application param), should this fall back to `window.location.origin` (i.e. back to homepage)?

ericbuckley replied:
We could be explicit about it here, but I believe that's what AuthApp does if the value is undefined/null/empty.  https://github.com/department-of-veterans-affairs/vets-website/blob/5b01d8ee1b1c70a9cf39f8fef8019d4fb6508930/src/applications/auth/containers/AuthApp.jsx#L131-L139

patrickvinograd replied:
Confirmed in localhost that passing a nonsense application name like "foobar" caused redirect to `http://localhost:3001/auth/login/callback/null`. I'm going to work on a fix

patrickvinograd reviewed (approved):
This looks a lot simpler, tracking this on the frontend. One comment about the fallback behavior for the externalRedirect case.

patrickvinograd commented on src/platform/user/authentication/utilities.js:
```diff
@@ -75,12 +68,25 @@ function redirectWithGAClientId(redirectUrl) {
   const returnUrl =
     window.location.pathname === '/sign-in/'
-      ? window.location.origin
+      ? standaloneRedirect()
```
```suggestion
      ? standaloneRedirect() || window.location.origin
```

patrickvinograd replied:
I tested above locally and it seemed to work. I think "null" isgetting written to sessionStorage and then appended to the base url in AuthApp.

ericbuckley replied:
thanks for testing and fixing

ericbuckley committed:
Update src/platform/user/authentication/utilities.js

Co-authored-by: Patrick Vinograd <patrick@adhocteam.us>

patrickvinograd committed:
Merge branch 'master' into eb-external-redirects

U-DON commented on src/platform/user/tests/authentication/utilities.unit.spec.js:
```diff
@@ -115,3 +105,33 @@ describe('authentication URL helpers', () => {
+
+  it('should return null when an application param is not provided', () => {
+    global.window.location.search = '';
+    expect(standaloneRedirect()).to.equal(null);
```
These might read better as:

```
expect(standaloneRedirect()).to.be.null;
```

patrickvinograd replied:
Fixed.

U-DON commented on src/platform/user/tests/authentication/utilities.unit.spec.js:
```diff
@@ -115,3 +105,33 @@ describe('authentication URL helpers', () => {
+    expect(standaloneRedirect()).to.equal(externalRedirects.myvahealth);
+  });
+
+  it('should strip any CRLF characters from the "to" parameter', () => {
```
How do CRLF characters end up in the parameter (a URL path)?

patrickvinograd replied:
This was implemented based on advice from: https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html

I.e. malicious entities injecting a URL parameter that leads to them being able to inject headers by splitting the URL in the HTTP request with a CRLF.

U-DON commented on src/platform/user/authentication/utilities.js:
```diff
@@ -75,12 +68,25 @@ function redirectWithGAClientId(redirectUrl) {
+
+  if (url && to) {
+    url = url.replace(new RegExp('/$'), '');
+    url += `/${to.replace(new RegExp('^/'), '').replace('\r\n', '')}`;
```
Could `trim()` be used instead for the CRLF?

U-DON replied:
Maybe more readable as:

```
let url;
const hostname = externalRedirects[application];

if (hostname) {
  const pathname = to.startsWith('/') ? to : `/${to}`;
  url = `${hostname}/${pathname};
  url.trim();
}
```

patrickvinograd replied:
The idea is that a malicious CRLF could appear anywhere, not just at the end of the URL.

U-DON replied:
Makes sense. Would rather see the validation on the backend, but in lieu of that, I would amend my suggestion to replace like the original:

```
url = `${baseUrl}${pathname}`.replace('\r\n', ''); // Prevent CRLF injection.
```

U-DON commented on src/platform/user/authentication/utilities.js:
```diff
@@ -75,12 +68,25 @@ function redirectWithGAClientId(redirectUrl) {
+  let url = externalRedirects[application] || null;
+
+  if (url && to) {
+    url = url.replace(new RegExp('/$'), '');
```
If we're going to cut off the trailing slash, do we need it in the first place in `externalRedirects`?

ericbuckley replied:
no we can remove it, the thought was to leave it here just in case someone later adds an externalRedirect with a trailing slash.

U-DON commented on src/platform/user/tests/authentication/utilities.unit.spec.js:
```diff
@@ -7,7 +7,10 @@ import {
+  standaloneRedirect,
+  externalRedirects,
 } from '../../authentication/utilities';
+import { executablePath } from 'puppeteer';
```
Was this meant to be added?

patrickvinograd replied:
Removed.

patrickvinograd committed:
PR feedback incorporated

ericbuckley committed:
case to auto redirect user to cerner

if the user is logged in with SSOe and they land on the stand alone sign
in page, we should redirect them to either the application they want to
go to (via the application query param they have set) or the home page

patrickvinograd reviewed (commented):
Latest sign-in redirect addition LGTM.

ericbuckley committed:
refactoring stadalone redirect to read better
